README.md
[![Travis Build Status](https://travis-ci.org/pneuvial/c3co.svg?branch=master)](https://travis-ci.org/pneuvial/c3co)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/pneuvial/c3co?branch=master&svg=true)](https://ci.appveyor.com/project/pneuvial/c3co)
[![Coverage Status](https://img.shields.io/codecov/c/github/pneuvial/c3co/master.svg)](https://codecov.io/github/pneuvial/c3co?branch=master)

# c3co: Inferring cancer cell clonality from copy number data


_WARNING 2019-03-03: The **c3co** package is under heavy development._


## Installation

```r
remotes::install_github("pneuvial/c3co", build_vignettes = TRUE)
```

## Model 

This package implements a constraint dictionary learning algorithm to recover subclones across several DNA copy number profiles. The c3co model is described in the figure below. Two heterogeneous tumor samples (green and yellow circles) are composed of a collection of normal cells (gray discs) and two cancer subclones (red triangles and blue squares). One of the cancer subclones is present in both tumor samples. 

![](vignettes/img/features.png)
![](vignettes/img/features2.png)

The corresponding (noiseless) total copy number profiles are displayed in the figure below. They are given by a linear combination of the latent profiles. This Figure is adapted from Nowak *et al.* (Biostatistics, 2011). 

![](vignettes/img/model.png)
![](vignettes/img/model2.png)

The `c3co` method aims at inferring the latent (blue, red and gray) copy-number profiles and the corresponding weights from the observed (yellow and green) copy-number profiles. Importantly, `c3co` takes advantage of the information of parental copy numbers, which are available from allelic ratios (a.k.a. B allele fractions or BAF) as measured by SNP array data or whole exome (WXS) or whole genome sequencing (WGS) data. 

## Documentation

See the package [vignette](vignettes/c3co.Rmd)


NEWS
Package: c3co
=============

Version: 0.4.0-9000 [2019-12-14]

NEW FEATURES:

 * c3co() is now a generic function with S3 methods for lists and NULL.


Version: 0.4.0 [2019-03-03]
 
SIGNIFICANT CHANGES:

 * Changed the default algorithm for limSolve::lsei() in internal get.W()
   from "lsei" (type=1) to "quadprog::solve.QP" (type=2).
   
 * positiveFusedLasso() now returns an NA model fit when the weight matrix
   becomes rank deficient in the C3CO algorithm.  Previously, it arbitrarily
   switched to fit one fewer subclone (K <- K-1).
   
 * positiveFusedLasso() and fitC3co() gained argument 'intercept' to control 
   whether a normal subclone should be modelled or not.

 * Renamed loadFacetsdata() and Facetswrapper() to loadFACETSdata() and
   FACETSwrapper().

SOFTWARE QUALITY:

* Improved documentation of mixSubClones() and buildSubclone().

* More sanity checks in mixSubClones() and get.W().
 
DEPRECATION & DEFUNCT:

 * Removed argument 'forceNormal' from initializeZt() because it was an ad-hoc
   approach to model a "normal" subclone, which is now taken care of by
   positiveFusedLasso(..., intercept = TRUE).


Version: 0.3.0 [2019-02-20]

SIGNIFICANT CHANGES:

 * Slots 'S' and 'S0' of 'posFused' were renamed to 'Zt' and 'Z0t'.
 
 * Renamed argument 'Z' of positiveFusedLasso() to 'Zt' to clarify that the
   input is the transpose of matrix 'Z'. Same for internal internal get.W().
   Analogously, argument 'Zhat' of modelFitStatistics() was renamed to 'Zhatt'.

 * Renamed initializeZ() to initializeZt() to clarify that it returns the
   transpose of matrix 'Z'.  Same for internal get.Z() renamed to get.Zt().

 * Renamed argument 'p' of initializeZt() to 'K'.

 * In getToyData(), replaced argument 'nbBkps' by 'nbSegs' (= nbBkps+1)
   for notation consistency.

 * mixSubclones() now requires that argument 'W' is matrix like, i.e. it
   has two dimensions.

NEW FEATURES:

 * Now positiveFusedLasso() records whether the model fit converged or not
   and how many iterations were performed.

 * Now positiveFusedLasso() produce an informative error message if attempted
   on a non-identifiable setup, i.e. more subclones (K) than samples (n).
   
 * loadFacetsdata() gained argument 'pattern'.

 * getToyData() gained arguments 'dimension' and 'returnLocus'.
 
BUG FIXES:

 * Argument 'lambda' of internal get.Zt() was neither invariant to the
   number of samples (n) nor the number of segments (J).
 
 * Internal get.W() would return a weight matrix list dimensions lost.

 * Passing argument 'Rowv' to heatmap.3() that is of class of 'dendrogram'
   would produce error 'length(x) = 2 > 1' in coercion to 'logical(1)' in
   R-devel with _R_CHECK_LENGTH_1_LOGIC2_=true.

 * Generic trackNames() called NextMethod() instead of UseMethod().
 
 * In getToyData(), robustified the creation of Z so that it is full rank.
 
 * In getToyData(), robustified the creation of W so all row sum to 1 and all
   items are in [0,1].
 
SOFTWARE QUALITY:

 * Add first tests for get.W(), get.Z() and positiveFusedLasso().

 * Add help pages with examples to internal get.W() and get.Z() functions.
 
 * Add sanity checks and tests for getLocusData(). 


Version: 0.2.6-9000 [2018-03-19]

NEW FEATURES:

 * Now loadPSCBSdata() and segmentData() preserve any sample names.
   
 * Add C3coSegmentation class with print() method etc.

 * Using message() everywhere (instead of print()).
 
 * Force normal component in Z.
 
 * Check rank of W.
 
 * segmentData() asserts that all input data use the same set of loci,
   there are no "missing" chromosomes, and that there are at least three
   data points per chromosome.

SOFTWARE QUALITY:

 * Added test scripts.
 

Version: 0.2.5-9000 [2017-07-30]

NEW FEATURES:

 * Added method modelFitStats() for objects of class 'posFused'.
 
 * Cleanups in fitC3co() (definition of BIC, consistency with log-likelihood).
   analysis. Added corresponding test script.
   
 * Added several flavors for the initialization of Z: svd, NMF, and archetypal 
   analysis. Added corresponding test script.
   
 * Added function getToyData() (and corresponding test script) to generate toy 
   locus-level and segment-level data.

SOFTWARE QUALITY:

 * Added simple test scripts to inst/testScripts.
 


Version: 0.2.4-9000 [2017-07-24]

SIGNIFICANT CHANGES:

 * In fitC3co(): the model is now applied on centered data.
 
NEW FEATURES:

 * buildSubclone() can now also generate purely simulated data if no real data
   are given as input.
   

Version: 0.2.3-9000 [2017-06-28]

SIGNIFICANT CHANGES:

 * Force Z1 <= Z2.
 
BUG FIXES:

 * Bugs fixed for W set type=2 in limSolve::lsei().


Version: 0.2.2-9000 [2017-05-17]

SIGNIFICANT CHANGES:

 * Changed structure of c3co object by adding 'config' slot.
 
BUG FIXES:

 * Bugs fixed for Z inference when (WtW)^-1 is non-inversible.
 

Version: 0.2.1-9000 [2017-04-20]

SIGNIFICANT CHANGES:

 * Removed conditions c1 and c2 in fitC3co().
 
 * Changed generation of weight matrix (new function rSparseWeightMatrix()).
 
 * Set lambda1=lambda2 for Minor and Major inference by default.


Version: 0.2.0-9000 [2017-04-12]

SIGNIFICANT CHANGES:

 * Changed default grid of values for penalization parameters: now ranges
   uniformly *on the log scale* over the interval.
   
SOFTWARE QUALITY:

 * More code cleanups.


Version: 0.2.0 [2017-03-31]

CODE REFACTORING:

 * Added function initializeZ() so that the initialization is
   performed only once per candidate number of subclones in 'c3co'.
   
 * In class 'posFused', added slot 'S0' to store initial estimates of Z
 
 * Moved function get.Z.new() into get.Z()
 
 * Added function fitC3co() that performs the actual fitting from segment-level
   data. This function is called by the main function c3co().

SOFTWARE QUALITY:

 * More code cleanups.
 
 * Added minimal sanity checks for some function arguments.
 

Version: 0.1.15 [2017-03-21]

NEW FEATURES:

 * createZdf() now handles multiple chromosomes and statistics.


R/c3co.R
#' Cancer subclone inference
#'
#' @param dat A list of data frames for each patient, each of them of the form:
#' \describe{
#'   \item{`tcn`}{Total copy number}
#'   \item{`dh`}{Mirrored B allele fraction}
#'   \item{`pos`}{Position on the genome}
#'   \item{`chr`}{Chromosome}
#' }
#'
## FIXME: What does ".. list composed of : the ..." mean? /HB 2018-03-04
#' @param parameters.grid Is a list composed of : the penalty coefficients
#' named either `lambda`, `lambda1` and `lambda2`. For the model on minor and
#' major CN, it is possible to give two grids (one for minor CN and one for
#' major CN), and all combination is tested. If only one grid is given the
#' consider  that `lambda1` = `lambda2`.
#' and composed of : a vector named `nb.arch` of integers which is the number
#' of features in the model.
#'
#' @param stat `"TCN"` or `"C1C2"`.
#'
#' @param segDat Either a path to the file that contains segmentation
#' (\file{*.rds} file), by default `NULL`, or a file that contains segmentation 
#' (e.g. from [segmentData()]).
#'
#' @param \dots Further arguments to be passed to [fitC3co()].
#'
#' @param verbose A logical value indicating whether to print extra
#' information. Defaults to `FALSE`.
#'
#' @return An object of class [c3coFit][c3coFit-class].
#'
#' @examples
#' dataAnnotTP <- acnr::loadCnRegionData(dataSet="GSE11976", tumorFrac=1)
#' dataAnnotN <- acnr::loadCnRegionData(dataSet="GSE11976", tumorFrac=0)
#' len <- 500*10   ## Number of loci
#' K <- 3L         ## Number of subclones
#' n <- 15L        ## Number of samples
#' set.seed(88)
#' bkps <- list(c(100, 250)*10, c(150, 400)*10, c(150, 400)*10)
#' regions <- list(c("(1,2)", "(0,2)", "(1,2)"),
#' c("(0,3)", "(0,1)", "(1,1)"), c("(0,2)", "(0,1)", "(1,1)"))
#' datSubClone <- buildSubclones(len=len, nbClones=K, bkps=bkps, regions=regions,
#'                               dataAnnotTP=dataAnnotTP, dataAnnotN=dataAnnotN)
#' W <- rSparseWeightMatrix(nb.samp=n, nb.arch=K, sparse.coeff=0.7)
#' dat <- mixSubclones(subClones=datSubClone, W=W)
#' l1 <- seq(from=1e-6, to=1e-4, length.out=10L)
#' parameters.grid <- list(lambda=l1, nb.arch=2:5)
#' res <- c3co(dat, parameters.grid=parameters.grid)
#' ## FIXME: BUG: propagate NA when rank deficient
#' ## FIXME: resC <- c3co(dat, stat="TCN", parameters.grid =parameters.grid)
#' @importFrom methods new
#' @export
c3co <- function(dat, ...) UseMethod("c3co")


#' @rdname c3co
#' @export
c3co.list <- function(dat, parameters.grid=NULL, stat=c("C1C2", "TCN"),
                 segDat=NULL, ..., verbose=FALSE) {
    ## Sanity checks
    stat <- match.arg(stat)
    if (!is.null(dat)) {
        if (!is.list(dat)) {
            stop("Argument 'dat' should be a list")
        }
        checkCols <- lapply(dat, FUN=function(dd) {
            coln <- colnames(dd)
            ecn <- c("tcn", "dh", "pos", "chr") ## expected
            mm <- match(ecn, table=coln)
            if (anyNA(mm)) {
                stop("Argument 'dat' should contain columns named ",
                     comma(sQuote(ecn)))
            }
        })
        expectedL <- nrow(dat[[1]])
        checklength <- lapply(dat, FUN=function(dd) {
            nrowDD <- nrow(dd)
            if (nrowDD != expectedL) {
                stop("Each data.frame in 'dat' should have the same size")
            }
        })
    }
    if (is.null(parameters.grid)) {
        lambda1 <- seq(from=1e-6, to=1e-4, length.out=10L)
        nb.arch <- 2:(length(dat)-1L)
        parameters.grid <- list(lambda=lambda1, nb.arch=nb.arch)
    }

    checkGrid <- lapply(names(parameters.grid), FUN=function(na) {
        ecn <- c("lambda", "lambda1", "lambda2", "nb.arch") ## expected
        mm <- match(na, table=ecn)
        if (anyNA(mm)) {
            stop("Argument 'parameters.grid' should contain ",
                 comma(sQuote(ecn)))
        }
    })

    if (!is.null(segDat)) {
      if (is.character(segDat)) {
        if (verbose) {
            message("Reading segmentation results from file: ")
            mprint(segDat)
        }
        seg <- readRDS(segDat)
      } else {
        ## Sanity check
        checkGrid <- lapply(names(segDat), FUN=function(na) {
          ecn <- c("bkp", "Y1", "Y2", "Y") ## expected
          mm <- match(na, table=ecn)
          if (anyNA(mm)) {
              stop("Argument 'parameters.grid' should contain ",
                   comma(sQuote(ecn)))
          }
        })
        message("Segmented data is provided, skip segment step")
        seg <- segDat
      }
    } else {
        seg <- segmentData(dat, stat=stat, verbose=verbose)
    }
    bkpList <- seg$bkp

    reslist <- new("c3coFit")
    reslist@bkp <- bkpList
    reslist@segDat <- list(Y1=seg$Y1, Y2=seg$Y2, Y=seg$Y)

    if (stat == "C1C2") {
        Y1 <- t(seg$Y1)
        Y2 <- t(seg$Y2)
    } else if (stat == "TCN") {
        Y1 <- t(seg$Y)
        Y2 <- NULL
    }

    fit <- fitC3co(Y1, Y2=Y2, parameters.grid=parameters.grid,
                           ..., verbose=verbose)
    reslist@fit <- fit$fit
    reslist@config <- fit$config
    
    reslist
}


#' @rdname c3co
#' @export
c3co.NULL <- function(dat, ...) {
  ## AD HOC: Why do we get c3co(NULL) and what is it really supposed
  ## to return. /HB 2019-05-31
  c3co.list(dat, ...)
}

